Add extension for NewRelic

goldenplec opened the pull request:
Creating a New Relic Layers for the Extra-PHP-Extensions Repository of Bref.

goldenplec commented:
Looks like all checks have passed which is great. So I'll await approval/merge. Let me know if there's any questions

NeilJ247 commented:
@goldenplec looks like you have some minor conflicts but hopefully this can get merged in soon :)

Nyholm commented on .travis.yml:
```diff
@@ -29,6 +29,7 @@ env:
   - LAYER=mailparse PHP="72 73 74"
   - LAYER=mongodb PHP="72 73 74"
   - LAYER=msgpack PHP="72 73 74"
+  - LAYER=newrelic PHP="72 73 74"
```
This is not needed any more. But can you update `.github/workflows/ci.yml`?

goldenplec replied:
Will do. This was in the documented steps which is why I did it.

Nyholm replied:
It was true at the time you made the PR. But I've been doing a lot of changes last 10 days.

goldenplec replied:
:) Not a problem happens from time to time and great to hear you've been adding stuff 👍

Nyholm commented on CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
 # Change log
 
+## 0.5.4
```
Should be 0.7.1

Nyholm commented on Readme.md:
```diff
@@ -68,6 +68,7 @@ memory_limit=128M
 | Memcached  | `${bref:extra.memcached-php-74}` | `extension=/opt/bref-extra/memcached.so`   |
 | MongoDB    | `${bref:extra.mongodb-php-74}`   | `extension=/opt/bref-extra/mongodb.so`     |
 | MsgPack    | `${bref:extra.msgpack-php-74}`   | `extension=/opt/bref-extra/msgpack.so`     |
+| Newrelic    | `${bref:extra.newrelic-php-74}`   | `extension=/opt/bref-extra/newrelic.so`     |
```
"`extension=/opt/bref-extra/newrelic.so`" should not be needed

goldenplec replied:
I mirrored the other entries in the table. Is there a reason to not fill it in on the newrelic row?

Nyholm replied:
Yes, because you add it in the layer. See https://github.com/brefphp/extra-php-extensions/pull/92#discussion_r534513241

Nyholm commented on layers/newrelic/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+# Copy the newrelic.so and php.ini file to the tmp folder
+
+RUN cp `php-config --extension-dir`/newrelic.so /tmp/newrelic.so
+RUN cp /opt/bref/etc/php/php.ini /tmp/php.ini
```
```suggestion
RUN echo 'extension=/opt/bref-extra/newrelic.so' > /tmp/ext.ini
```

goldenplec replied:
I'm confused by this. php.ini is where the .so is loaded. What is ext.ini in this context and the extension= is already achieved in the Dockerfile inside the php.ini with the configuration settings.

Nyholm commented on layers/newrelic/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+  /tmp/newrelic-php5-*/newrelic-install install && \
+  rm -rf /tmp/newrelic-php5-* /tmp/nrinstall* 
+  
+RUN echo $' \n \
+extension=newrelic.so \n \
+newrelic.appname = ${NEW_RELIC_APP_NAME} \n \
+newrelic.license = ${NEW_RELIC_LICENSE_KEY} \n \
+newrelic.daemon.address = ${NEW_RELIC_DAEMON_ADDRESS} \n \
+newrelic.logfile = ${NEW_RELIC_LOG_FILE} \n \
+newrelic.loglevel = ${NEW_RELIC_LOG_LEVEL} \n \
+newrelic.daemon.app_connect_timeout = ${NEW_RELIC_DAEMON_APP_CONNECT_TIMEOUT} \n \
+newrelic.daemon.start_timeout = ${NEW_RELIC_DAEMON_START_TIMEOUT} \n \
+' >> /opt/bref/etc/php/php.ini
```
These are things specific to the application. It is the user that should add these. Right?

I suggest removing these.

goldenplec replied:
To work, the agent must be capable of connecting to a remote running daemon. 
so 
extension=newrelic.so (needed to tell PHP to load the newrelic.so file)
newrelic.appname = ${NEW_RELIC_APP_NAME}.   (Needs to be set or the Lambda will default report to PHP Application)
newrelic.license = ${NEW_RELIC_LICENSE_KEY}   (Needed to report to your account)
newrelic.daemon.address = ${NEW_RELIC_DAEMON_ADDRESS}  (Needed to tell the agent where the remote daemon is located)

The following configuration settings are definitely options but having control of them via the Lambda Environment variables is much more scalable and you can then have your settings like key and daemon address stores in something like KMS if you wanted. For example, using Serverless we defined ENV VARS and those ENV VARS become the settings below. In particular it's useful to set log file to STDERR location so it will log to Cloudwatch.

newrelic.logfile = ${NEW_RELIC_LOG_FILE} \n \
newrelic.loglevel = ${NEW_RELIC_LOG_LEVEL} \n \
newrelic.daemon.app_connect_timeout = ${NEW_RELIC_DAEMON_APP_CONNECT_TIMEOUT} \n \
newrelic.daemon.start_timeout = ${NEW_RELIC_DAEMON_START_TIMEOUT} \n \

I can remove them, but I'm interested in where in the Lambda you would propose the end user sets these values. PHP will only read the php.ini, a file in the Scan Directory or .user.ini in the local folder. Maybe I'm oblivious to something obvious but I only believe I can amend php.ini or anything in the scan dir. So unless I use .user.ini is your suggestion that the user of this extension should put it in their PHP code using **ini_set** ?

Nyholm replied:
Yes, I suggest that we dont add any config here. We should let the default values provided by newrelic take effect. 

In your application you should add `php/conf.d/mysettings.ini` where you write: 

```
newrelic.appname = ${NEW_RELIC_APP_NAME}
newrelic.license = ${NEW_RELIC_LICENSE_KEY}
newrelic.daemon.address = ${NEW_RELIC_DAEMON_ADDRESS}
newrelic.logfile = ${NEW_RELIC_LOG_FILE}
newrelic.loglevel = ${NEW_RELIC_LOG_LEVEL}
newrelic.daemon.app_connect_timeout = ${NEW_RELIC_DAEMON_APP_CONNECT_TIMEOUT}
newrelic.daemon.start_timeout = ${NEW_RELIC_DAEMON_START_TIMEOUT} 
```

Another user may specify hard coded values here instead of the environment variables.

goldenplec replied:
We can make that change, the agent won't work at all without a license key and remote daemon_address though so anyone who implements it will need to know to add a mysettings.ini file. 

If we remove these settings, then we can remove the line copying the php.ini to the extension layer. The only change from the core php.ini to the one in my layer is these config changes. 

So I can make the container literally just install New Relic and copy over the .so file to where it needs to be. Then add additional documentation on implementation of configuration.

Nyholm commented on layers/newrelic/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+# Copy the two key files to the correct location for the empty layer.
+# newrelic.so should become a layered extension. php.ini should overwrite base layer
+COPY --from=ext /tmp/newrelic.so /opt/bref/lib/php/extensions/no-debug-zts-20190902/newrelic.so
+COPY --from=ext /tmp/php.ini /opt/bref/etc/php/php.ini
```
```suggestion
COPY --from=ext /tmp/ext.ini /opt/bref/etc/php/conf.d/ext-newrelic.ini

```

goldenplec replied:
I can't find reference of this ext.ini and where it comes from or is used. Can you provide context or a docs link?

Nyholm replied:
Have a look at the Ldap layer: https://github.com/brefphp/extra-php-extensions/blob/0.8.1/layers/ldap/Dockerfile#L14-L22

Line 21: Write the file `/opt/bref-extra/ldap.so`
Line 15: Add `extension=/opt/bref-extra/ldap.so` to a temp file `/tmp/ext.ini`
Line 22: Move the temp file to a path that PHP reads config from

goldenplec replied:
So the New Relic installation creates a newrelic.ini in the Scan Dir. However the Scan Dir isn't being read. Which is why I was appending to php.ini for simplicity. I'll create a newrelic.ini in the same dir as php.ini as it seems from your suggestion this should work too.

Nyholm commented on layers/newrelic/README.md:
```diff
@@ -0,0 +1,56 @@
+
+### Extra configuration files
+
+The configuration in the PHP.ini for New Relic is purposely using Environment Variables. In your Serverless config you can simple add them manually or store them in something like KMS and reference them as required if the setting is a token/key/password.
```
Does a php.ini file really read environment variables like this?

Hm.. I still thing you should specify that in your app. You can still use environment variables if you want to.

goldenplec replied:
Can you provide an example. I elaborated above on what these are being used for but the ENV VARS seemed a scalable way to provide unique configuration per Lambda and control the settings when deploying quickly and easily. Totally open to switching, just wondering what the preferred approach is

Nyholm reviewed (changes requested):
Thank you. 

A lot of changes has happen in the past week. 

Could you rebase this and look at some other layers to add a `config.json` and `test.php`?

goldenplec commented:
@Nyholm I believe I've rebased to all suggestions. Let me know if I missed anything.

tlfbrito commented:
You should remove the `.DS_Store`

tip: Add to the ignore list of your global git config file

goldenplec commented:
@tifbrito - Thanks didn't realise I had included it. My bad using --all and trusting I didn't generate those files 👍 

I've rebased again since the changes on the main repo had newer changes again so I'm hoping this time we may have the success :)

goldenplec commented:
Actually I'm seeing now there's test file and config.json that's needed so I need to do another push.

tlfbrito commented on Readme.md:
```diff
@@ -58,10 +58,11 @@ memory_limit=128M
 | Memcached          | `${bref:extra.memcached-php-74}`      |
 | MongoDB            | `${bref:extra.mongodb-php-74}`        |
 | MsgPack            | `${bref:extra.msgpack-php-74}`        |
+| Newrelic           | `${bref:extra.newrelic-php-74}	`      |
```
remove extra space to align `|` (OCD review :P)

goldenplec replied:
Don't worry about reviewing for the moment. Something in the newest bref layer confuses the PHP Agent installer into thinking it's ZTS compiled and our installer is putting the ZTS version of our agent in instead of the non-ZTS.

I can manually pull the non-ZTS version but I want to see if there's a way to over-ride the installer to avoid having to put checks of PHP API into the Dockerfile if it's not neccessary. I'm hoping to figure this out over the coming days and get another push to you.

goldenplec replied:
OK I've resolved all that. Build looked good locally now. Tested with a basic app again and working and reporting fine. So hopefully this will be the last rebase. Rebased as changelog 0.9.4 config and test file are included this time. Sorry for the back and forth on this one.

tlfbrito commented:
pinging @mnapoli and @Nyholm to give a check on this.

Nyholm commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
-.php_cs.cache
\ No newline at end of file
+.php_cs.cache
+.DS_Store
+/layers/newrelic/.DS_Store
```
These should not be here. 

You should add `.DS_Store` in your global gitignore.

Nyholm commented on layers/newrelic/README.md:
```diff
@@ -0,0 +1,85 @@
+# Bref New Relic Extension
```
Let's put this file in the /doc directory

goldenplec replied:
No problem. Moving now and resubmitting.

Nyholm commented on layers/newrelic/README.md:
```diff
@@ -0,0 +1,85 @@
+
+We would recommend pointing the logfile to `/proc/self/fd/2` which is STDERR and you will therefore get logging to Cloudwatch in the event of an issue. Log Level defaults to INFO but if you only want issues and not conversational logging you can raise the level to `warning` or `error`. Daemon address should be the IP of the location of your remotely started daemon (see below for guidance.)
+
+If you wish you use ENV VARS to achieve this. You can set the values to 
```
Excellent

goldenplec replied:
I've reworked this due to the confusion @mnapoli had with it. Hoping the reworked is better.

mnapoli commented on layers/newrelic/README.md:
````diff
@@ -0,0 +1,85 @@
+    NEW_RELIC_DAEMON_START_TIMEOUT: "5s"You may add extra 
+```   
+
+We would recommend pointing the logfile to `/proc/self/fd/2` which is STDERR and you will therefore get logging to Cloudwatch in the event of an issue. Log Level defaults to INFO but if you only want issues and not conversational logging you can raise the level to `warning` or `error`. Daemon address should be the IP of the location of your remotely started daemon (see below for guidance.)
````
Are you sure about that?

I tried again (because I did already a year ago) and it doesn't seem possible to write to that file inside lambda:

```php
    $stderr = fopen('/proc/self/fd/2', 'ab');
    fwrite($stderr, 'This is a test log into stderr');
    fclose($stderr);
```

<img width="1151" alt="Screen 20201221 2ZFSpHLy@2x" src="https://user-images.githubusercontent.com/720328/102808539-be466b80-43c0-11eb-8c02-d652e877ae79.png">

goldenplec replied:
It's logged to Cloudwatch logs. So I can see what the PHP agent is trying to do using that. Here's an example. Without it if you have issues it'll be hard to see what the agent was trying to do. Lambda being read only this is the only approach I found that has worked.

<img width="1555" alt="Screenshot 2020-12-23 at 00 45 54" src="https://user-images.githubusercontent.com/28544519/102946357-4ce9e400-44b8-11eb-9a62-356e13f500e7.png">

mnapoli replied:
I'm not sure how it works for the daemon and not the lambda function, but if it works, it works! Thanks!

mnapoli commented on layers/newrelic/README.md:
````diff
@@ -0,0 +1,85 @@
+    NEW_RELIC_LOG_LEVEL: "info"
+    NEW_RELIC_DAEMON_APP_CONNECT_TIMEOUT: "5s"
+    NEW_RELIC_DAEMON_START_TIMEOUT: "5s"You may add extra 
+```   
````
I'm not sure I understand, the paragraph above mentions `php/conf.d/newrelic-settings.ini` and the one above environment variables. We should make that a bit easier.

How about documenting only environment variables? In that case let's also make the example above clearer:

```yaml
```suggestion
provider:
    environment:
        NEW_RELIC_APP_NAME: "BrefApp"
        ...
```
```

goldenplec replied:
I removed ENV VARS from the only approach in the doc on the suggestion of @Nyholm 

i'll just add a bit more clarity to the README.

mnapoli replied:
Yes, to be honest at the moment the `Extra configuration files` looks confusing because it mixes ini config, environment variables and serverless.yml, without being clear what the order is to change all these files.

Maybe something simpler like:

1. edit serverless.yml
2. edit the ini config file

goldenplec replied:
The ini file isn't extra or at this stage optional. You need to set a license key. I was hard adding these values in the php.ini file but it was recommend by @Nyholm to move these out and let the end user create the config and control it in their code. 

So the extra detail provided in the newrelic-readme should hopefully now explain it out. With the Env Vars options being a way of controlling them from your serverless config or via a KMS but you would also still have the newrelic-settings.ini file, as it will have the values needed, the readme now explains the default values too so if a user wants to they can simply have a newrelic-settings.ini file with just daemon address and license key and it would work. The other values are the config values I'd determined have the most relevance for someone using the PHP agent in this Lambda environment. There's a lot more config options but they aren't needed. Hope that helps unravel some of it.

mnapoli commented on layers/newrelic/README.md:
```diff
@@ -0,0 +1,85 @@
+
+In this example, the New Relic Daemon will be running on ECS or an EC2 instance listening on port 31339. If you follow the New Relic documentation and just install the daemon to a host, you can have it listen for incoming connectivity.
+
+If you wish to run the daemon as a container. This repo should be particularly useful.
```
How about recommending to use https://github.com/newrelic/serverless-newrelic-lambda-layers?

That sounds perfect for Lambda, no need for a separate EC2 or container, and there's a serverless plugin that hides all the complexity.

goldenplec replied:
Running the daemon in a layer as part of the same function could potentially have race conditions as every cold start the daemon is going to have to start. This is why I'm recommending the daemon run independently. 

The daemon is not a feature of Python or Node. Those are singular in terms of agent approach. PHP agent is in two pieces. Agent and Daemon. So the serverless layers are perfect for the likes of Node and Python but for PHP it's the potential of lost data.

mnapoli reviewed (commented):
Really happy to see this moving forward!

Nyholm commented on layers/newrelic/README.md:
```diff
@@ -0,0 +1,85 @@
+    NEW_RELIC_LOG_FILE: "/proc/self/fd/2"
+    NEW_RELIC_LOG_LEVEL: "info"
+    NEW_RELIC_DAEMON_APP_CONNECT_TIMEOUT: "5s"
+    NEW_RELIC_DAEMON_START_TIMEOUT: "5s"You may add extra 
```
```suggestion
    NEW_RELIC_DAEMON_START_TIMEOUT: "5s"
```

goldenplec replied:
Removed

mnapoli replied:
I can still see the old line though.

goldenplec replied:
Odd I removed that and pushed the change. Let me double check the fork and my local code again.

goldenplec commented on Readme.md:
```diff
@@ -58,10 +58,11 @@ memory_limit=128M
 | Memcached          | `${bref:extra.memcached-php-74}`      |
 | MongoDB            | `${bref:extra.mongodb-php-74}`        |
 | MsgPack            | `${bref:extra.msgpack-php-74}`        |
+| Newrelic           | `${bref:extra.newrelic-php-74}	`     |
```
Nope that's fair, frustrating that slipped in.

mnapoli commented on layers/newrelic/README.md:
```diff
@@ -0,0 +1,85 @@
+    NEW_RELIC_LOG_FILE: "/proc/self/fd/2"
+    NEW_RELIC_LOG_LEVEL: "info"
+    NEW_RELIC_DAEMON_APP_CONNECT_TIMEOUT: "5s"
+    NEW_RELIC_DAEMON_START_TIMEOUT: "5s"You may add extra 
```
Why should we override these default values?

For every optional ini setting that we override there should be a good reason, else this is just noise for users I think.

For example, why do we change the log level? Why do we change the timeouts?

I do understand for `NEW_RELIC_LOG_FILE` and the other settings above though, because this is required to work on Lambda. But for the other settings I don't understand.

goldenplec commented:
@mnapoli - Thanks for the back and forth on the above. 

Moved the Readme to the Docs folder
Updated the Readme to attempt to be more clear about the config and the option to use ENV VARS/KMS if wanted.

Hopefully this reads a bit better, sorry I'm not a great docs writer :)

goldenplec commented:
@mnapoli @Nyholm just wanted to check if this might get merged?

Nyholm commented:
I’m all good. I’ll merge it this weekend if nobody says no.

Nyholm commented:
Thank you

elazar commented:
@Nyholm Will a tagged release with this addition be out soon?

mnapoli commented:
Thanks for merging! With the holidays I wasn't able to keep up, the new README looks good, thanks!

Nyholm commented:
Cool cool cool. 

Thank you. 
This was released as 0.9.4

kristijorgji commented:
Can you provide a working example of this extension?

I added it under the layers

```
        layers:
            - ${bref-extra:newrelic-php-82}
```

How do I specify the newrelic info like appname licence etc now ?

mnapoli commented:
@kristijorgji please read the docs of this project.

kristijorgji commented:
@mnapoli I read the entire docs, also tried them but with no result.

I posted the full question and also what I tried here
https://stackoverflow.com/questions/78852240/how-to-use-php-bref-newrelic-extension-for-aws-lambda

Would be very grateful for some help to have bref newrelic lambda extension working

Thanks

Nyholm commented:
Hello. 

Have read of the documentation: https://github.com/brefphp/extra-php-extensions/blob/master/docs/newrelic.md
The first code example is very much what you need. 


Also, consider starting a new issue so we dont ping people on a 2 year old thread.
